Move asset installation to the ModChooser

penev92 opened the pull request:
Closes #8990.

That ticket pretty much explains what is going on here and why.

The progress bar will probably need polishing. Other than that I think it's done.

I **strongly** suggest reviewing the code by commits and not the entire diff.

penev92 committed:
Move ContentInstaller to OpenRA.Game from Mods.Common

penev92 committed:
Add mod asset checks to ModBrowserLogic

penev92 committed:
Remove asset handling from BlankLoadScreen

penev92 committed:
Add paths to TestFiles

File checks for existing files don't work otherwise.

penev92 committed:
Move D2k tilesets to the root asset directory

penev92 committed:
Enable mod asset package downloading from the modchooser

penev92 committed:
Enable mod asset installation from CD from the modchooser

penev92 committed:
Add content installation dialogs to `modchooser`

penev92 committed:
Pipe mod music installation through the ModChooser

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -159,5 +194,41 @@ void LoadMod(ModMetadata mod)
+			var filesList = new List<string>();
+			foreach (var folder in folders)
+			{
+				var path = Platform.ResolvePath(folder).Replace("/", "\\");
```
This hard-codes the directory separator to the Windows only `\` which breaks everywhere else so please change this to
```csharp
var path = Platform.ResolvePath(folder);
```

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -159,5 +194,41 @@ void LoadMod(ModMetadata mod)
+				if (!Directory.Exists(path))
+					continue;
+
+				var filesInDir = Directory.GetFiles(path).Select(f => f.ToLower());
```
This isn't just the files, but the file + folder. It is `~/.openra/Content/` on Unix so making this `~/.openra/content/` will break on case sensitive file systems. Change it to:

```csharp
var filesInDir = Directory.GetFiles(path);
```

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -159,5 +194,41 @@ void LoadMod(ModMetadata mod)
+
+			var files = filesList.ToHashSet();
+
+			return mod.AssetInformation.TestFiles.All(f => files.Contains(Platform.ResolvePath(f).Replace("/", "\\").ToLower()));
```
This combines both mistakes of above and breaks the detection on non-Windows systems completely. The following will work:

```charp
return mod.AssetInformation.TestFiles.All(f => files.Contains(Platform.ResolvePath(f)));
```

penev92 commented:
Fixed the path check workarounds.

Mailaender commented on OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackagesLogic.cs:
```diff
@@ -52,9 +57,9 @@ void ShowDownloadDialog()
 			var cancelButton = panel.Get<ButtonWidget>("CANCEL_BUTTON");
 
-			var mirrorsFile = Platform.ResolvePath("^", "Content", Game.ModData.Manifest.Mod.Id, "mirrors.txt");
+			var mirrorsFile = Platform.ResolvePath("^", "Content", "mirrors.txt");
```
We probably want to store this in `Path.GetTempPath()` instead.

pchote replied:
It can have a `Path.GetRandomFileName()` too (and don't forget to remove it when you're done).

pchote replied:
Or even better, we could go with my suggestion (which was ignored) from the original PR, and use the version of the `Download` constructor that returns a string instead of saving a file to disk.

pchote replied:
See `MapCache.QueryRemoteMapDetails` for an example on how to use that version.

pchote replied:
You missed this in the last pass.

Mailaender commented:
From a usability standpoint if I run the game with `OpenRA.Game.exe Game.Mod=ts` it will now start the `modchooser` and highlights the first entry (which is `d2k`). As we store `Game.PreviousMod` in the settings we could highlight it in the mod chooser in that case so you immediately see the install assets button of the mod you wanted to start.

Mailaender commented:
Content detection and download works fine on Linux. :white_check_mark:

LipkeGu commented:
@penev92 you could also apply the TFD changes when you are on it so it is done in one way

pchote commented on OpenRA.Game/ContentInstaller.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace OpenRA
+{
+	public class ContentInstaller : IGlobalModData
```
Could you please add a comment `// Referenced from ModMetadata, so needs to be in OpenRA.Game :(` before this?

pchote commented on OpenRA.Game/ModMetadata.cs:
```diff
@@ -24,6 +24,7 @@ public class ModMetadata
 		public string Version;
 		public string Author;
 		public bool Hidden;
+		public ContentInstaller AssetInformation;
```
Can we call the variable `Installer` or something else more closely related to the type?

pchote commented on OpenRA.Mods.Common/InstallUtils.cs:
```diff
@@ -93,6 +68,10 @@ public static string GetMountedDisk(Func<string, bool> isValidDisk)
 
+					var dir = Path.GetDirectoryName(dest);
+					if (!Directory.Exists(dir))
+						Directory.CreateDirectory(dir);
```
It looks like you can now remove the `Directory.CreateDirectory(destPath);` at the start of the method.

LipkeGu replied:
Line 72 can also be removed...

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackagesLogic.cs:
```diff
@@ -21,24 +20,30 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 			progressBar = panel.Get<ProgressBarWidget>("PROGRESS_BAR");
 			statusLabel = panel.Get<LabelWidget>("STATUS_LABEL");
 
+			var modName = ModMetadata.AllMods[modId].Title;
```
You could improve separation by passing the modName (`modTitle`?) instead of the modId and doing this lookup.

pchote replied:
Oh, nevermind.  I see that you use `modId` in other places too.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Installation/InstallFromCDLogic.cs:
```diff
@@ -74,6 +76,9 @@ void CheckForDisk()
 					InstallTFD(Platform.ResolvePath(path, "data1.hdr"));
 			else
 			{
+				var info = insertDiskContainer.Get<LabelWidget>("INFO2");
```
`INFO2` isn't a great ID.  Can we call this `MOD_TITLE` instead?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -140,18 +147,46 @@ void RebuildModList()
-		void SelectMod(ModMetadata mod)
+		void SelectMod(ModMetadata mod, bool? isInstalled)
 		{
+			if (isInstalled == null)
```
This is pretty ugly.  You can make this cleaner and also save some unnecessary checks when scrolling through mods by defining a `Cache<string, bool> modInstallStatus`.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -140,18 +147,46 @@ void RebuildModList()
+		void LoadMod(ModMetadata mod, bool? isInstalled)
 		{
+			if (isInstalled == null)
+				isInstalled = IsModInstalled(mod);
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -140,18 +147,46 @@ void RebuildModList()
+					{ "modId", mod.Id }
+				};
+
+				modChooserPanel.Get<BackgroundWidget>("bg").Visible = false;
```
Widget IDs in yaml should be in upper case.

pchote replied:
You can / should adjust the mod chooser to open these as a Window so that you don't need to manually hide/show them.

pchote commented on mods/cnc/chrome.yaml:
```diff
@@ -146,31 +145,6 @@ textfield-focused: chrome.png
 	corner-bl: 64,254,2,2
 	corner-br: 126,254,2,2
 
-# A copy of button
```
Is this deliberate?  Progress bars are used in more places than just the installer!

Mailaender replied:
I checked the map download but it vanishes so fast that I am not sure.

pchote replied:
You can confirm that it is used there by checking the lobby yaml.

penev92 replied:
Yeah, didn't think of map download. Will bring it back.

pchote commented on mods/d2k/chrome.yaml:
```diff
@@ -173,30 +173,6 @@ scrollbar: buttons.png
 	up_arrow: 32,140,16,16
 	up_pressed: 32,140,16,16
 
-# A copy of dialog3 (pressed button)
```
Likewise here.

pchote commented on OpenRA.Game/ContentInstaller.cs:
```diff
@@ -0,0 +1,39 @@
+{
+	public class ContentInstaller : IGlobalModData
+	{
+		public readonly string MenuWidget = null;
```
These three `*Widget` variables are now unused.

pchote commented on mods/ra/chrome.yaml:
```diff
@@ -502,30 +502,6 @@ scrollbar: buttons.png
 	up_arrow: 132,140,16,16
 	up_pressed: 132,140,16,16
 
-# A copy of dialog3 (pressed button)
```
Here too.

pchote commented:
The buttons in the main mod chooser UI are 35px high.  All your installer buttons are 32px high.

pchote commented:
After the asset download completes it takes you back to the mod chooser.  It should try to load the mod instead (which can then drop back to the chooser itself if something is busted).

pchote commented:
The "Back" button from the installation panel will reload the mod chooser.  That should no longer be necessary.

pchote commented:
I volunteer to use a YAML polish pass, so you can focus on the code nits for now.

penev92 committed:
Fix ModChooser UI

penev92 committed:
Remove obsolete installation UI YAML files from the official mods

penev92 committed:
Streamline mod changing after installation

 - Don't return to the ModChooser after a successful installation.
 - Don't reload the ModChooser after a cancelled installation.

penev92 commented:
Rebased and addressed all requests, minus the button sizes.

pchote commented on OpenRA.Game/ModMetadata.cs:
```diff
@@ -46,6 +47,9 @@ public class ModMetadata
 				mod.Id = m;
 
+				if (nd.ContainsKey("ContentInstaller"))
+					mod.Content = FieldLoader.Load<ContentInstaller>(nd["ContentInstaller"]);
```
It might pay to wrap this in a try/catch.  You don't want to kill the whole game if someone has one bogus mod installed.

penev92 replied:
Actually the `if` should take care of that. I also just reverted locally the addition of `ContentInstaller:` to `modchooser\mod.yaml`, because it's no longer needed (was needed in the earlier stages of this branch).

pchote replied:
The `ContentInstaller` block may exist, but have bogus content that makes FieldLoader throw.

pchote commented on OpenRA.Mods.Common/LoadScreens/BlankLoadScreen.cs:
```diff
@@ -40,27 +40,13 @@ public void StartGame(Arguments args)
 			var installData = Game.ModData.Manifest.Get<ContentInstaller>();
-			var installModContent = !installData.TestFiles.All(f => GlobalFileSystem.Exists(f));
-			var installModMusic = args != null && args.Contains("Install.Music");
+			var isModContentInstalled = installData.TestFiles.All(f => GlobalFileSystem.Exists(f.Substring(f.LastIndexOf('/') + 1)));
```
`Path.GetFileName`?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Installation/DownloadPackagesLogic.cs:
```diff
@@ -21,24 +20,30 @@ namespace OpenRA.Mods.Common.Widgets.Logic
 
+			var modName = ModMetadata.AllMods[modId].Title;
+			var text = panel.Get<LabelWidget>("TITLE").Text;
+			panel.Get<LabelWidget>("TITLE").Text = text.F(modName);
```
This adds an edge case that will complicate the future translation string extraction (if this ever happens).  Can you hardcode the format string in the logic file instead for now? (this is what we do everywhere else).

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Installation/InstallLogic.cs:
```diff
@@ -17,31 +15,28 @@ namespace OpenRA.Mods.Common.Widgets.Logic
-				Ui.OpenWindow("INSTALL_DOWNLOAD_PANEL", widgetArgs);
+			var mod = ModMetadata.AllMods[modId];
+			var text = panel.Get<LabelWidget>("DESC1").Text;
+			panel.Get<LabelWidget>("DESC1").Text = text.F(mod.Title);
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -148,10 +155,29 @@ void SelectMod(ModMetadata mod)
 			if (selectedIndex - modOffset > 4)
 				modOffset = selectedIndex - 4;
+
+			loadButton.Text = modInstallStatus[mod] ? "Load Mod" : "Install Assets";
```
The screen that pops up after "Install Assets" says "Game assets not found!". This wording assumed that we kept the "Load Mod" button text.  Can you please either keep the button text, or change the title of the dialog?

pchote replied:
My preference is to change the dialog title to "Install Assets", which then matches "Install Music".

pchote commented on OpenRA.Mods.Common/Widgets/Logic/ModBrowserLogic.cs:
```diff
@@ -148,10 +155,29 @@ void SelectMod(ModMetadata mod)
 		{
+			if (!modInstallStatus[mod])
+			{
+				var contentData = mod.Content;
```
Just use `mod.Content` directly below.

pchote commented:
A handful of UI nits:
* Please reduce the height of the buttons on the main mod chooser panel so that they are 32px everywhere.
* The progress bar bar should be pure grey, without the inner red border (I can help with art).
* The layout of the download (and I assume also the copy from CD) dialog is bogus.  I suggest making the progress bar thinner to give the extra space you need (see point above).
* Please increase the height of all of these dialogs (keeping them equal) so that there is an equal margin to the right and below the buttons.
* Please take 4d740e897bf1f6ec05093eb221ffd6ad9f8d923e and then set an appropriate BaseLine to center the text in the mod chooser buttons.
* You can take 2affae44beccb4852351be0bcbb5c66a13d19b47 to fix the broken overlaps that happen at low percentages.


![screen shot 2015-09-08 at 21 24 04](https://cloud.githubusercontent.com/assets/167819/9746561/8328ef02-5670-11e5-910e-60af832cc2c7.png)

LipkeGu commented:
#9136 was also a try to make the progressbars more easier to configure :)

pchote commented:
b25aafa96d766109355468e0f45d257d7b7593bd gives you a new progress bar to play with.

![screen shot 2015-09-08 at 22 22 15](https://cloud.githubusercontent.com/assets/167819/9747851/4e3b6d08-5678-11e5-8b7b-44ff9499dbe0.png)

penev92 committed:
Add ButtonBaseLine metric

penev92 committed:
Generalise progress bars

penev92 commented:
That should be all now.

pchote commented:
There are a few more nits:

* Use the download ctor that doesn't need a temporary on-disk file (see discussion above).
* You forgot to set ButtonBaseLine: 1 for the modchooser mod.
* The download panel layout is still off: move the text down by a few px to match the second line in the other panels, and then center the progress bar in the gap.
* The button margins and the horizontal rule margin on the main browser panel don't match the dialogs.  Fixing this means tweaking the whole layout of the browser panel, so I won't insist on fixing that now.

penev92 commented:
Changed as requested.

Mailaender commented:
You added new stylecop errors.

pchote commented on mods/modchooser/metrics.yaml:
```diff
@@ -3,10 +3,12 @@
 	ButtonTextColorDisabled: 128,128,128
 	ButtonTextContrast: false
 	ButtonTextContrastColor: 0,0,0
+	ButtonBaseLine: 0
```
This resets the BaseLine back to 0!

pchote commented:
Confirmed that my requests above have been fixed, except for the `ButtonBaseLine` (again!).  :+1: after fixing that and the stylecop errors.  Good work!

penev92 commented:
OK they should be centered now.

pchote commented:
Good job :+1:

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/1442fb34743e0196e966b725f10a680261512f80
